admin/add_copyright.py

```diff
@@ -0,0 +1,124 @@
+# Copyright (c) 2013 Sandia Corporation.
+# This software is distributed under the BSD License.
+# Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+# the U.S. Government retains certain rights in this software.
+# For more information, see the README.rst file.
+#
+# ****************************************************************************
+
+import os
+import re
+import subprocess
+
+import version
+
+REGEX_EMAIL = re.compile(r"\d+\+([^@]+)@users\.noreply\.github\.com")
+HEADER_END = (
+    " ****************************************************************************"
+)
+
+
+def getLicenseSection(filename, comment_char="#"):
+    date = subprocess.check_output(
+        ["git", "log", "-1", "--format=format:%ad", "--date=short", "--", filename],
+        universal_newlines=True,
+    )
+    year = date[:4]
+    yield comment_char + "  ************************** Copyrights and license ***************************"
+    yield comment_char
+    yield comment_char + " This file is part of gcovr {}, a parsing and reporting tool for gcov.".format(
+        version.__version__
+    )
+    yield comment_char + " https://gcovr.com/en/stable"
+    yield comment_char
+    yield comment_char + " _____________________________________________________________________________"
+    yield comment_char
+    yield comment_char + " Copyright (c) " + year + " the gcovr authors"
+    yield comment_char + " Copyright (c) 2013 Sandia Corporation."
+    yield comment_char + " This software is distributed under the BSD License."
+    yield comment_char + " Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,"
+    yield comment_char + " the U.S. Government retains certain rights in this software."
+    yield comment_char + " For more information, see the README.rst file."
+    yield comment_char
+    yield comment_char + HEADER_END
+
+
+def addCopyrightToPythonFile(filename, lines):
+    # Empty file should be kept empty
+    if len(lines) == 0:
+        return lines
+
+    newLines = []
+    # Keep the Shebang
+    if lines[0].startswith("#!"):
+        newLines.append(lines.pop(0))
+        newLines.append("")
+
+    # Add encoding
+    newLines.append("# -*- coding:utf-8 -*-")
+    newLines.append("")
+
+    # Add license information
+    for line in getLicenseSection(filename):
+        newLines.append(line)
+
+    # Skip header and add rest of file
+    headerEndReached = False
+    skipEmptyLines = True
+    for line in lines:
+        if not headerEndReached:
+            if len(line) > 0 and line[0] == "#" and line[1:] == HEADER_END:
+                headerEndReached = True
+                continue
+
+        if headerEndReached:
+            if skipEmptyLines:
+                if len(line) == 0:
+                    continue
+                skipEmptyLines = False
+                newLines.append("")
+            newLines.append(line)
+
+    if not headerEndReached:
+        newLines.extend(lines)
```

Comment:
This kind of `headerEndReached` and `skipEmptyLines` state can be replaced by control flow and iterators. Whether to do that is a stylistic decision. I find control flow easier to follow than state machines.

```python
lines = iter(lines)

# skip lines until header end marker
for line in lines:
  if len(line) > 0 and line == '#' + HEADER_END:
    break

# skip empty lines, keeping the first non-empty line
for line in lines:
  if line:
    newlines.append("")
    newlines.append(line)
    break

# keep all other lines
newlines.extend(lines)
```

Reply:
I've never used an iterator explicite but this is a good approach. The only missing thing is if the header isn't found. I'll use this.